Document middleware usage (#138)
vercel/swr-site: pages/docs/middlewares.en-US.md

import Callout from 'nextra-theme-docs/callout'

# Middlewares

<Callout emoji="✅">
  Please update to the latest version (≥ 1.0.0) to use this option.
</Callout>

The option `middlewares` is a new addition in SWR 1.0. It lets you execute code before and after SWR hooks, giving users the ability to attach processors to SWR for extending its functionality. 

## Usage

### API

> A middleware can access to the previous wrapped or original `useSWR` hook, and return a function with arguments of `useSWR`. Expect to return the transformed data with the composition of hook, arguments and your optional operations here.

Reviewer: ```suggestion
Middleware receive the return value of a `useSWR` hook and pass it along to the next middleware, if any. You can transform the data, extend the return value with new attributes, and perform any additional operations before or after the hook runs.
```

Reviewer 2: This is much better, thx!

```jsx
const middleware = (useSWRNext) => (key, fetcher, config) => {

>   // enter phase ...

Reviewer: ```suggestion
  // Before hook runs...
```
"enter phase" sounds like a bit too much jargon? Before/after is clear to me

  const swr = useSWRNext(key, fetcher, config)

>   // exit phase ...

Reviewer: ```suggestion
  // After hook runs...
```

  return swr
}
```

> Grab your middlewares into an array then pass to useSWR

Reviewer: ```suggestion
Pass middleware as an array to any `useSWR` hook.
```

## Example

### Request Logger

> Let's build a simple request logger middleware as example. Then use it to print out all the fetcher requests sending from SWR hook.

Reviewer: ```suggestion
Let's build a simple request logger middleware as an example. It simply prints out all the fetcher requests sent from the SWR hook.
```

```jsx
import useSWR from 'swr'

const fetcher (key) => fetch(`/api?key=${key}`).then(res => res.text())

const loggerMiddleware = (useSWRNext) => (key, fetcher, config) => {
  console.log(`Request to /api with param key=${key}`)
  return useSWRNext(key, fetcher, config)
}

useSWR(key, fetcher, {
  middlewares: [loggerMiddleware]
})
```

> Output in console

Reviewer: ```suggestion
Every time the SWR hook runs, it will output to the console:
```

```
Request to /api with param key=key1
Request to /api with param key=key2
Request to /api with param key=key3
```

### Nested Middlewares

> Middlwares are applied like layers of onions as the common practice of JavaScript community. 

Reviewer: ```suggestion
Middlwares are applied on top of each other, as commonly found in other libraries. If you pass multiple middleware to one hook:
```

> If you have multiple fetcher inside one hook

Reviewer: ```suggestion
```

Combined with paragraph above

```jsx
useSWR(key, fetcher, {
  middlewares: [A, B, C]
})
```

> The execute order will be `A -> B -> C` like below:

Reviewer: ```suggestion
The order of execution will be `A → B → C`, as shown below:
```

```
enter A
  enter B
    enter C
    exit  C
  exit  B
exit  A
```

> If you have any middleware configured in `SWRConfig`, it will obey the order from most outside to most inside. The order will be `A -> B`

Reviewer: We don't document that you can pass middleware to SWRConfig anywhere yet

Reviewer: ```suggestion
If you have any middleware configured in `SWRConfig`, the outermost middleware will run first. In the following example, the order will be `A → B`:
```

```jsx
<SWRConfig value={{ middleware: [myMiddleware] }}>

// or...

useSWR(key, fetcher, { middleware: [myMiddleware] })
```

### Extend

Middleware will be extended like regular options. For example:

```jsx
function Bar () {
  useSWR(key, fetcher, { middleware: [c] })
  // ...
}

function Foo() {
  return (
    <SWRConfig value={{ middleware: [a] }}>
      <SWRConfig value={{ middleware: [b] }}>
        <Bar/>
      </SWRConfig>
    </SWRConfig>
  )
}
```

is equivalent to:

```js
useSWR(key, fetcher, { middleware: [a, b, c] })
```

### Multiple Middleware

Each middleware wraps the next middleware, and the last one just wraps the SWR hook. For example:

```jsx
useSWR(key, fetcher, { middleware: [a, b, c] })
```

The order of middleware executions will be `a → b → c`, as shown below:

```
enter A
  enter B
    enter C
      useSWR()
    exit  C
  exit  B
exit  A
```

## Examples

### Request Logger

Let's build a simple request logger middleware as an example. It simply prints out all the fetcher requests sent from the SWR hook.

```jsx
import useSWR from 'swr'

const fetcher (key) => fetch(`/api?key=${key}`).then(res => res.text())

> const loggerMiddleware = (useSWRNext) => (key, fetcher, config) => {

Author: Better to keep consistence to either use named functions or arrow functions. The former one if you feel middleware function name might be import (like for devtools)

Reviewer 2: I think it's easier for people who are not an JS expert to understand this:

```js
function A() {
  return () => {}
}
```

Than:

```js
const A = () => () => {}
```
